Add a mz_internal.mz_sleep(N) function

antifuchs opened the pull request:
This function has side-effects, and behaves slightly differently from
pg_sleep(N): Where the postgres function returns a `void` type, this
function just returns the time at which sleeping finished.

That saves us from implementing evaluation / compilation rules for
void types, and weirdly enough it doesn't seem require us to declare
a function purely-side-effect-y to prevent it being optimized away.

Fixes: #5894

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/6043)
<!-- Reviewable:end -->

antifuchs commented:
Speculatively adding @benesch since you expressed opinions on the Issue; I suspect this PR may be too naive (it definitely is lazy in materialized views, but for interactive/test usage, it looks OK?)

benesch commented on src/sql/src/func.rs:
```diff
@@ -1988,6 +1988,9 @@ lazy_static! {
                 }), oid::FUNC_UNNEST_LIST_OID;
+            },
+            "mz_sleep" => Scalar {
+                params!(Float64) => UnaryFunc::Sleep, 2626;
```
Can you add this to the `MZ_INTERNAL_BUILTINS` list below instead? I think it would be best if we really discourage users from calling this.

benesch replied:
Also, looks like 2626 is the OID for `pg_sleep`, but I think we should allocate a new OID (as we do with e.g. `FUNC_MZ_ALL_OID`) since this function has different semantics and a different name from `pg_sleep`!

benesch commented on doc/user/data/sql_funcs.yml:
```diff
@@ -356,6 +356,13 @@
+      Sleeps during evaluation of the expression for the specified duration in seconds and returns
+      the time at which sleeping finished.  The mz_sleep function is intended to be called
+      interactively, or in transactions during testing. If called in a materialized view, it will be
+      optimized away as aggressively as possible.
```
I think I'd actual prefer if we didn't document this! I can't think of a situation where it would be useful to the user and explaining precisely _what_ it blocks while sleeping is tricky.

benesch reviewed (commented):
Thanks very much! Some comments within.

philip-stoev commented:
Thank you, this is much appreciated. Curiously, multiple workers working on different rows of the same query can sleep concurrently! Life is weird like that.

philip-stoev commented on doc/user/data/sql_funcs.yml:
```diff
@@ -356,6 +356,13 @@
     description: 'The `timestamp with time zone` representing when the query was executed.<br/><br/>**NOTE**:
       Users cannot define views with queries containing `now()`.'
 
+  - signature: mz_sleep(d: double precision) -> timestamptz
```
For some reason this is not a valid yaml:

```
11:29:41 PM: Building sites … ERROR 2021/03/09 21:29:41 failed to load data: failed to load data: "/opt/build/repo/doc/user/data/sql_funcs.yml:359:1": failed to unmarshal YAML: yaml: line 359: mapping values are not allowed in this context
```

antifuchs commented:
OK, updated with the review suggestions:
* Put it on the `mz_internal` catalog as `mz_internal.mz_sleep`
* Use an OID in the 13000s
* Removed docs.

benesch reviewed (approved):
Looks great to me!
